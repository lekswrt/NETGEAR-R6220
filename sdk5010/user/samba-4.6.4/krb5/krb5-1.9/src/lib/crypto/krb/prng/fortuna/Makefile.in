mydir=lib$(S)crypto$(S)krb$(S)prng$(S)fortuna
BUILDTOP=$(REL)..$(S)..$(S)..$(S)..$(S)..
LOCALINCLUDES = -I$(srcdir)/.. -I$(srcdir)/../.. \
		-I$(srcdir)/../../../$(CRYPTO_IMPL)		\
		-I$(srcdir)/../../../$(CRYPTO_IMPL)/sha2 	\
		-I$(srcdir)/../../../$(CRYPTO_IMPL)/aes 	\
		-I$(srcdir)/../../../$(CRYPTO_IMPL)/enc_provider
DEFS=
RUN_SETUP = @KRB5_RUN_ENV@
PROG_LIBPATH=-L$(TOPLIBD)
PROG_RPATH=$(KRB5_LIBDIR)

STLIBOBJS= \
	prng_fortuna.o entropy.o
OBJS= \
	$(OUTPRE)prng_fortuna.$(OBJEXT) \
	$(OUTPRE)entropy.$(OBJEXT)

SRCS=\
	$(srcdir)/entropy.c \
	$(srcdir)/prng_fortuna.c 

all-unix:: all-libobjs

includes:: depend

depend:: $(SRCS)

t_fortuna: t_fortuna.$(OBJEXT) $(SUPPORT_DEPLIB)
	$(CC_LINK) -o t_fortuna t_fortuna.$(OBJEXT) -lcom_err $(SUPPORT_LIB) $(CRYPTO_DEPLIB)

t_fortuna_make_oct: t_fortuna_make_oct.$(OBJEXT) $(SUPPORT_DEPLIB)
	$(CC_LINK) -o t_fortuna_make_oct t_fortuna_make_oct.$(OBJEXT) -lcom_err $(SUPPORT_LIB) $(CRYPTO_DEPLIB)

check-unix:: t_fortuna t_fortuna_make_oct
# ifdef TEST_FORTUNA
ifeq ("@PRNG_FORTUNA_TEST@","yes")
	$(RUN_SETUP) $(VALGRIND) ./t_fortuna_make_oct > t_fortuna_make_oct.result && \
	diff t_fortuna_make_oct.result t_fortuna_make_oct.expected
else
	$(RUN_SETUP) $(VALGRIND) ./t_fortuna
	$(RUN_SETUP) $(VALGRIND) ./t_fortuna_make_oct > t_fortuna_make_oct.result
endif

clean::
	$(RM) t_fortuna$(EXEEXT) t_fortuna.$(OBJEXT)  t_fortuna_make_oct.result  t_fortuna_make_oct$(EXEEXT)  t_fortuna_make_oct.$(OBJEXT)


clean-unix:: clean-libobjs

@lib_frag@
@libobj_frag@

