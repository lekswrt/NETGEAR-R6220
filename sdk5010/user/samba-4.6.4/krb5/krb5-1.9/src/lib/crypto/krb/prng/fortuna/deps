# 
# Generated makefile dependencies follow.
#
prng_fortuna.so prng_fortuna.po $(OUTPRE)prng_fortuna.$(OBJEXT): $(BUILDTOP)/include/autoconf.h \
  $(BUILDTOP)/include/krb5/krb5.h $(BUILDTOP)/include/osconf.h \
  $(BUILDTOP)/include/profile.h $(COM_ERR_DEPS) $(srcdir)/../../../builtin/aes/aes.h \
  $(srcdir)/../prng.h \
  $(srcdir)/../../../builtin/sha2/sha2.h \
  $(top_srcdir)/include/k5-buf.h \
  $(top_srcdir)/include/k5-err.h $(top_srcdir)/include/k5-gmt_mktime.h \
  $(top_srcdir)/include/k5-int-pkinit.h $(top_srcdir)/include/k5-int.h \
  $(top_srcdir)/include/k5-platform.h $(top_srcdir)/include/k5-plugin.h \
  $(top_srcdir)/include/k5-thread.h $(top_srcdir)/include/k5-trace.h \
  $(top_srcdir)/include/krb5.h $(top_srcdir)/include/krb5/authdata_plugin.h \
  $(top_srcdir)/include/krb5/locate_plugin.h $(top_srcdir)/include/krb5/preauth_plugin.h \
  $(top_srcdir)/include/port-sockets.h $(top_srcdir)/include/socket-utils.h \
  prng_fortuna.c fortuna.h 
entropy.so entropy.po $(OUTPRE)entropy.$(OBJEXT): $(BUILDTOP)/include/autoconf.h \
  $(BUILDTOP)/include/krb5/krb5.h $(BUILDTOP)/include/osconf.h \
  $(BUILDTOP)/include/profile.h $(COM_ERR_DEPS)	\
  $(top_srcdir)/include/k5-buf.h \
  $(srcdir)/../prng.h \
  $(top_srcdir)/include/k5-err.h $(top_srcdir)/include/k5-gmt_mktime.h \
  $(top_srcdir)/include/k5-int-pkinit.h $(top_srcdir)/include/k5-int.h \
  $(top_srcdir)/include/k5-platform.h $(top_srcdir)/include/k5-plugin.h \
  $(top_srcdir)/include/k5-thread.h $(top_srcdir)/include/k5-trace.h \
  $(top_srcdir)/include/krb5.h $(top_srcdir)/include/krb5/authdata_plugin.h \
  $(top_srcdir)/include/krb5/locate_plugin.h $(top_srcdir)/include/krb5/preauth_plugin.h \
  $(top_srcdir)/include/port-sockets.h $(top_srcdir)/include/socket-utils.h \
  entropy.c
t_fortuna.so t_fortuna.po $(OUTPRE)t_fortuna.$(OBJEXT): $(BUILDTOP)/include/autoconf.h \
  $(BUILDTOP)/include/krb5/krb5.h $(BUILDTOP)/include/osconf.h \
  $(BUILDTOP)/include/profile.h $(COM_ERR_DEPS) $(srcdir)/../../../builtin/aes/aes.h \
  $(srcdir)/../prng.h \
  $(srcdir)/../../../builtin/sha2/sha2.h \
  $(top_srcdir)/include/k5-buf.h \
  $(top_srcdir)/include/k5-err.h $(top_srcdir)/include/k5-gmt_mktime.h \
  $(top_srcdir)/include/k5-int-pkinit.h $(top_srcdir)/include/k5-int.h \
  $(top_srcdir)/include/k5-platform.h $(top_srcdir)/include/k5-plugin.h \
  $(top_srcdir)/include/k5-thread.h $(top_srcdir)/include/k5-trace.h \
  $(top_srcdir)/include/krb5.h $(top_srcdir)/include/krb5/authdata_plugin.h \
  $(top_srcdir)/include/krb5/locate_plugin.h $(top_srcdir)/include/krb5/preauth_plugin.h \
  $(top_srcdir)/include/port-sockets.h $(top_srcdir)/include/socket-utils.h \
  t_fortuna.c fortuna.h 

