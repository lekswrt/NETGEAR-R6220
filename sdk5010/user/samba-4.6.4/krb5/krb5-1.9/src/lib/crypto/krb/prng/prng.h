/* -*- mode: c; c-basic-offset: 4; indent-tabs-mode: nil -*- */
/*
 * lib/crypto/krb/prng/prng.h
 *
 * Copyright 2010 by the Massachusetts Institute of Technology.
 * All Rights Reserved.
 *
 * Export of this software from the United States of America may
 *   require a specific license from the United States Government.
 *   It is the responsibility of any person or organization contemplating
 *   export to obtain such a license before exporting.
 *
 * WITHIN THAT CONSTRAINT, permission to use, copy, modify, and
 * distribute this software and its documentation for any purpose and
 * without fee is hereby granted, provided that the above copyright
 * notice appear in all copies and that both that copyright notice and
 * this permission notice appear in supporting documentation, and that
 * the name of M.I.T. not be used in advertising or publicity pertaining
 * to distribution of the software without specific, written prior
 * permission.  Furthermore if you modify this software you must label
 * your software as modified software and not distribute it in such a
 * fashion that it might be confused with the original M.I.T. software.
 * M.I.T. makes no representations about the suitability of
 * this software for any purpose.  It is provided "as is" without express
 * or implied warranty.
 */


#ifndef PRNG_H
#define PRNG_H

#include "k5-int.h"

#if defined(FORTUNA)
#define ENTROPY_BUFSIZE 32  /* SHA256 digest length */
#elif defined(CRYPTO_IMPL_NSS)
/*
 * NSS gathers its own OS entropy, so it doesn't really matter how much we read
 * in krb5_c_random_os_entropy.  Use the same value as Yarrow (without using a
 * Yarrow constant), so that we don't read too much from /dev/random.
 */
#define ENTROPY_BUFSIZE 20
#else
#define ENTROPY_BUFSIZE YARROW_SLOW_THRESH/8  /* SHA1 digest length*/
#endif

/* prng.h */
struct krb5_prng_provider {
    char prng_name[8];
    krb5_error_code (*make_octets)(krb5_context, krb5_data *);
    krb5_error_code (*add_entropy)(krb5_context, unsigned int randsource, const krb5_data*);
    int (*init)(void);
    void (*cleanup)(void);
};

#endif
