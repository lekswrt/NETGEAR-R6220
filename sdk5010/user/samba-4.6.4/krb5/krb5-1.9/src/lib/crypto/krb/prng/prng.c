/* -*- mode: c; c-basic-offset: 4; indent-tabs-mode: nil -*- */
/*
 * Copyright (C) 2001, 2002, 2004, 2007, 2008, 2010 by the Massachusetts Institute of Technology.
 * All rights reserved.
 *
 *
 * Export of this software from the United States of America may require
 * a specific license from the United States Government.  It is the
 * responsibility of any person or organization contemplating export to
 * obtain such a license before exporting.
 *
 * WITHIN THAT CONSTRAINT, permission to use, copy, modify, and
 * distribute this software and its documentation for any purpose and
 * without fee is hereby granted, provided that the above copyright
 * notice appear in all copies and that both that copyright notice and
 * this permission notice appear in supporting documentation, and that
 * the name of M.I.T. not be used in advertising or publicity pertaining
 * to distribution of the software without specific, written prior
 * permission.  Furthermore if you modify this software you must label
 * your software as modified software and not distribute it in such a
 * fashion that it might be confused with the original M.I.T. software.
 * M.I.T. makes no representations about the suitability of
 * this software for any purpose.  It is provided "as is" without express
 * or implied warranty.
 */

#include "prng.h"

#ifdef FORTUNA
#include "fortuna.h"
const struct krb5_prng_provider *prng = &krb5int_prng_fortuna;
#elif defined(CRYPTO_IMPL_NSS)
#include "prng_nss.h"
const struct krb5_prng_provider *prng = &krb5int_prng_nss;
#else
#include "yarrow.h"
const struct krb5_prng_provider *prng = &krb5int_prng_yarrow;
#endif

/*
 * krb5int_prng_init - Returns 0 on success
 */
int krb5int_prng_init(void)
{
    int err = 0;
    err = prng->init();
    return err;
}

/*
 * krb5_c_random_add_entropy - Returns 0 on success
 */
krb5_error_code KRB5_CALLCONV
krb5_c_random_add_entropy(krb5_context context, unsigned int randsource,
                          const krb5_data *data)
{
    krb5_error_code err = 0;
    err = prng->add_entropy(context, randsource, data);
    return err;
}

/*
 * krb5_c_random_seed - Returns 0 on success
 */
krb5_error_code KRB5_CALLCONV
krb5_c_random_seed(krb5_context context, krb5_data *data)
{
    return krb5_c_random_add_entropy(context, KRB5_C_RANDSOURCE_OLDAPI, data);
}

/*
 * krb5_c_random_make_octets -  Returns 0 on success
 */
krb5_error_code KRB5_CALLCONV
krb5_c_random_make_octets(krb5_context context, krb5_data *data)
{
    krb5_error_code err = 0;
    err = prng->make_octets(context, data);
    return err;
}

void
krb5int_prng_cleanup (void)
{
    prng->cleanup();
    return;
}


/*
 * Routines to get entropy from the OS.  For UNIX we try /dev/urandom
 * and /dev/random.  Currently we don't do anything for Windows.
 */
#if defined(_WIN32)

krb5_error_code KRB5_CALLCONV
krb5_c_random_os_entropy(krb5_context context, int strong, int *success)
{
    if (success)
        *success = 0;
    return 0;
}

#else /*Windows*/
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_SYS_STAT_H
#include <sys/stat.h>
#endif

/*
 * Helper function to read entropy from  a random device.  Takes the
 * name of a device, opens it, makes sure it is a device and if so,
 * reads entropy.  Returns  a boolean indicating whether entropy was
 * read.
 */

/* 
 * read_entropy_from_device - Returns 0 on success
 */
static int
read_entropy_from_device(krb5_context context, const char *device)
{
    krb5_data data;
    struct stat sb;
    int fd;
    unsigned char buf[ENTROPY_BUFSIZE], *bp;
    int left;
    fd = open (device, O_RDONLY);
    if (fd == -1)
        return 0;
    set_cloexec_fd(fd);
    if (fstat(fd, &sb) == -1 || S_ISREG(sb.st_mode)) {
        close(fd);
        return 0;
    }

    for (bp = buf, left = sizeof(buf); left > 0;) {
        ssize_t count;
        count = read(fd, bp, (unsigned) left);
        if (count <= 0) {
            close(fd);
            return 0;
        }
        left -= count;
        bp += count;
    }
    close(fd);
    data.length = sizeof (buf);
    data.data = (char *) buf;
    return (krb5_c_random_add_entropy(context, KRB5_C_RANDSOURCE_OSRAND,
                                      &data) == 0);
}

/* 
 * krb5_c_random_os_entropy - Returns 0 on success
 */
krb5_error_code KRB5_CALLCONV
krb5_c_random_os_entropy(krb5_context context, int strong, int *success)
{
    int unused;
    int *oursuccess = success ? success : &unused;

    *oursuccess = 0;
    /* If we are getting strong data then try that first.  We are
       guaranteed to cause a reseed of some kind if strong is true and
       we have both /dev/random and /dev/urandom.  We want the strong
       data included in the reseed so we get it first.*/
    if (strong) {
        if (read_entropy_from_device(context, "/dev/random"))
            *oursuccess = 1;
    }
    if (read_entropy_from_device(context, "/dev/urandom"))
        *oursuccess = 1;
    return 0;
}

#endif /*Windows or pre-OSX Mac*/

