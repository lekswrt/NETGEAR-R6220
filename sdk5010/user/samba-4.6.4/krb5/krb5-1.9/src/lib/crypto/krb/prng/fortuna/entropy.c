/* -*- mode: c; c-basic-offset: 4; indent-tabs-mode: nil -*- */
/*
 * lib/crypto/krb/prng/fortuna/entropy.c
 *
 * Copyright 2010 by the Massachusetts Institute of Technology.
 * All Rights Reserved.
 *
 * Export of this software from the United States of America may
 *   require a specific license from the United States Government.
 *   It is the responsibility of any person or organization contemplating
 *   export to obtain such a license before exporting.
 *
 * WITHIN THAT CONSTRAINT, permission to use, copy, modify, and
 * distribute this software and its documentation for any purpose and
 * without fee is hereby granted, provided that the above copyright
 * notice appear in all copies and that both that copyright notice and
 * this permission notice appear in supporting documentation, and that
 * the name of M.I.T. not be used in advertising or publicity pertaining
 * to distribution of the software without specific, written prior
 * permission.  Furthermore if you modify this software you must label
 * your software as modified software and not distribute it in such a
 * fashion that it might be confused with the original M.I.T. software.
 * M.I.T. makes no representations about the suitability of
 * this software for any purpose.  It is provided "as is" without express
 * or implied warranty.
 */

/* various methods to collect entropy */

#include "prng.h"

#include "fortuna.h"
#include "k5-int.h"

#ifndef min
#define min(a, b)       ((a) < (b) ? (a) : (b))
#endif

krb5_error_code
k5_entropy_from_device(krb5_context context, const char *device, unsigned char* buf, int buflen)
{
    struct stat sb;
    int fd;
    unsigned char *bp;
    size_t left;
    fd = open(device, O_RDONLY);
    if (fd == -1)
        return 0;
    set_cloexec_fd(fd);
    if (fstat(fd, &sb) == -1 || S_ISREG(sb.st_mode)) {
        close(fd);
        return 0;
    }

    for (bp = buf, left = sizeof(buf); left > 0;) {
        ssize_t count;
        count = read(fd, bp, (unsigned) left);
        if (count <= 0) {
            close(fd);
            return 0;
        }
        left -= count;
        bp += count;
    }
    close(fd);
    return 0;
}

krb5_error_code
k5_entropy_dev_random(krb5_context context, unsigned char* buf, int buflen)
{
    memset(buf, 0, buflen);
    return k5_entropy_from_device(context,"/dev/random", buf, buflen);
}

krb5_error_code
k5_entropy_dev_urandom(krb5_context context, unsigned char* buf, int buflen)
{
    memset(buf, 0, buflen);
    return k5_entropy_from_device(context,"/dev/urandom", buf, buflen);
}

krb5_error_code
k5_entropy_pid(krb5_context context, unsigned char* buf, int buflen)
{
    pid_t pid = getpid(); 
    int pidlen = min(buflen,(int)sizeof(&pid));
    memset(buf, 0, buflen);
    memcpy(buf, &pid, pidlen);
    return 0;
}

krb5_error_code
k5_entropy_uid(krb5_context context, unsigned char* buf, int buflen)
{
    pid_t uid = getuid(); 
    int uidlen=min(buflen,(int)sizeof(&uid));
    memset(buf, 0, buflen);
    memcpy(buf, &uid, uidlen);
    return 0;
}

#ifdef TEST_FORTUNA
int
test_entr(krb5_context context, unsigned char* buf, int buflen)
{
    char buf1[26] = "Seed To Test Fortuna PRNG";
    memset(buf, 0, buflen);
    memcpy(buf, buf1, min(buflen, 26));
    return 0;
}
#endif
