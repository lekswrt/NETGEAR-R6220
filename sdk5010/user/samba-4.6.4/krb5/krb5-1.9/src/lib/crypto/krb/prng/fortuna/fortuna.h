/* -*- mode: c; c-basic-offset: 4; indent-tabs-mode: nil -*- */
/*
 * lib/crypto/krb/prng/fortuna/fortuna.h
 *
 * Copyright 2010 by the Massachusetts Institute of Technology.
 * All Rights Reserved.
 *
 * Export of this software from the United States of America may
 *   require a specific license from the United States Government.
 *   It is the responsibility of any person or organization contemplating
 *   export to obtain such a license before exporting.
 *
 * WITHIN THAT CONSTRAINT, permission to use, copy, modify, and
 * distribute this software and its documentation for any purpose and
 * without fee is hereby granted, provided that the above copyright
 * notice appear in all copies and that both that copyright notice and
 * this permission notice appear in supporting documentation, and that
 * the name of M.I.T. not be used in advertising or publicity pertaining
 * to distribution of the software without specific, written prior
 * permission.  Furthermore if you modify this software you must label
 * your software as modified software and not distribute it in such a
 * fashion that it might be confused with the original M.I.T. software.
 * M.I.T. makes no representations about the suitability of
 * this software for any purpose.  It is provided "as is" without express
 * or implied warranty.
 */

#ifndef FORTUNA_H
#define FORTUNA_H

#include "k5-int.h"
#include "prng.h"
#ifndef OPENSSL
#include "aes.h"
#endif
#include "enc_provider.h"
#include "sha2.h"
#include "enc_provider.h"

extern const struct krb5_prng_provider krb5int_prng_fortuna;

/* various entropy collect functions */
krb5_error_code
k5_entropy_from_device(krb5_context context, const char *device, unsigned char* buf, int buflen);
krb5_error_code
k5_entropy_dev_random(krb5_context context, unsigned char* buf, int buflen);
krb5_error_code
k5_entropy_dev_urandom(krb5_context context, unsigned char* buf, int buflen);
krb5_error_code
k5_entropy_pid(krb5_context context, unsigned char* buf, int buflen);
krb5_error_code
k5_entropy_uid(krb5_context context, unsigned char* buf, int buflen);

#ifdef TEST_FORTUNA
int test_entr(krb5_context context, unsigned char* buf, int buflen);
#endif

#define FORTUNA_OK                1  /* All is well */
#define FORTUNA_FAIL              0  /* generic failure */
#define FORTUNA_LOCKING          -12

#endif
