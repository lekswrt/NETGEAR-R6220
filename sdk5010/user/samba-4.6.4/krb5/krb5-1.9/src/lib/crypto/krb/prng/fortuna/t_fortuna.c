/*
 * Copyright (c) 2007 Kungliga Tekniska HÃ¶gskolan
 * (Royal Institute of Technology, Stockholm, Sweden).
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the Institute nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "prng.h"
#include "fortuna.h"
#include <k5-int.h>


#define LEN_TEST_BUF 1024 * 1024
static int len = LEN_TEST_BUF;

int
main(int argc, char **argv)
{
    char buffer[LEN_TEST_BUF];
    krb5_data data = {0, LEN_TEST_BUF, (char*)buffer};
    int bit, i;
    double res;
    int bits[8] = { 0, 0, 0, 0, 0, 0, 0, 0 };

    if (len < 100000) 
        return 0;

    for (i = 0; i < LEN_TEST_BUF; i++)
         buffer[i] = 0;

    /* head vs tail */
    krb5_c_random_make_octets(NULL, &data);
    for (i = 0; i < len; i++) {
        unsigned char c = ((unsigned char *)buffer)[i];
        for (bit = 0; bit < 8 && c; bit++) {
            if (c & 1)
                bits[bit]++;
            c = c >> 1;
        }
    }
    
    for (bit = 0; bit < 8; bit++) {

        res = ((double)abs(len - bits[bit] * 2)) / (double)len;
        if (res > 0.005){
            printf("head %d vs tail %d > 0.5%%%% %lf == %d vs %d\n",
                 bit, bit, res, len, bits[bit]);
            return 1;
        }

        printf("head vs tails bit %d is %lf\n", bit, res);
    }

    return 0;
}
