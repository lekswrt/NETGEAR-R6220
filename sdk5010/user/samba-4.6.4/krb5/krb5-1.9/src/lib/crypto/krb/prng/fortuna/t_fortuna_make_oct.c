/*
 * Copyright (c) 2007 Kungliga Tekniska HÃ¶gskolan
 * (Royal Institute of Technology, Stockholm, Sweden).
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the Institute nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "prng.h"
#include "fortuna.h"
#include <k5-int.h>

#define LEN_TEST_BUF 1024
static int len = LEN_TEST_BUF;

static void hex_print( FILE* f, const char* var, void* data, size_t size );

int
main(int argc, char **argv)
{
    char buffer[LEN_TEST_BUF];
    krb5_data data = {0, LEN_TEST_BUF, (char*)buffer};
    int i;

    for (i = 0; i < LEN_TEST_BUF; i++)
         buffer[i] = 0;

    krb5_c_random_make_octets(NULL, &data);

    hex_print( stdout, "random1", data.data, data.length );

    /* To target FORTUNA_RESEED_BYTE */
    i = 0;
    while (i++ < 11){
        krb5_c_random_make_octets(NULL, &data);
    }

    hex_print( stdout, "random2", data.data, data.length );

    return 0;
}
static void
hex_print( FILE* f, const char* var, void* data, size_t size )
{
    const char* conv = "0123456789abcdef";
    size_t i;
    char* p = (char*) data;
    char c, d;

    fprintf( f, var );
    fprintf( f, " = " );
    for ( i = 0; i < size; i++ )
    {
        c = conv[ (p[ i ] >> 4) & 0xf ];
        d = conv[ p[ i ] & 0xf ];
        fprintf( f, "%c%c", c, d );
    }
    fprintf( f, "\n" );
}

