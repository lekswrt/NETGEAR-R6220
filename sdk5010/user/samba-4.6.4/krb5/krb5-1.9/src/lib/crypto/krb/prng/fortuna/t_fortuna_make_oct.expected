random1 = c27d031fce09ad26525e107021c1038062c9e05672065bce1139a9ae583d9cd70319e9a68d1f4da2cf24099af05ab1a910c781035864ab7e45337199cfa7ece05111d3604215feac555e847148dcf47fa8361746ce162201edbb83f4f753ad9933f9c618f86e15ff28eecdddbcaacd2ff212c236141b2edc5e5b91f2a95f89af5070eed9652f2857963801cff9f21d6a08f8c1480270009ebde555a00dcc60b0688cba89950aeeac8a0806922ea1baae494e1bc1be7920fa6c9b980a80f374a5ea53f3e10ffeef047bea388949fc0b66252d25c8775dd01a03a27528a4c801993129e581e3c4d77ecca9efaf27047d7eec025f6f0601d469a93f5be5cbac10b773a392a3d0a167ec76637014974fe6808e88b9b8fbd51dfbad7192d1edad971b11404f96849cc88dab07b21ff14d5d528d60c3bb7ad4b1ddc3e6b082df9d89b020d276db81676ffcf6521b2bab33c404b27f15f5547ee4a37654680fb331afb256a6cd83146a5baf70a6a5a455c7ead49493f7aa85d4afde951f33f44a99c49ea1e3f568b70735c46dce20460150d2c96aa0849cd5b7062253fdb11ba4a7e046fe24ccd3c87bbebe3351c10a8b6fb4b1e2dd46803782982970cf9805fa2a8540d9f910ea4e360a342a320cce34101472fdc258ba339038e464da6bc20b79df5fc6ae27762a7315d3b342d99eca5d23c5ce2f1ba800bdb68882a6b520e31037e8c1e1ea3444334ed6ec3913994344b714009b9cc816d57d5a74e1cbfe3f038f0def48f43ef81983a872068f60e808ed75997d23e4896faba5542738bd403b3d5fc33f64d606d376e936f89a60cf0ad417ba0f9cd4bb70287437e9f7d63572432d859473dadd9b8c51a8813e3e734968288afdb22a5779c1f6ccc681cf2332f634096f4d40088bdee97270a7d7c04d7b367001b7c154da3f69734fd91590710e54af97f4e03b80c800ccef3282b519aaa4a76f9cb06d6746211be6ca8ac910b3b0bb799ae185dbc92c8ac2d91f0c6e5bc7095e41ebd2b3217648ba62616a3820f282154ab2787f26c47c47bb099810b65e862ed44b9d47ac87d693d2f87295834274310f337b15a4d7b33d91ffefa7b3dfec270327db4b46a5e0e1cdc072e5d0a2e8392a2bede194ee715070785f1d1cb6a5bcd2f13b4aa644541b9e6336bbb87d88bead71858dfc4c211589865faf08c8cc9ba47728dc8d2ebc3fb6d9c434b8ad7f99fbc14124fb5ffaa24fb0c7650bb19e5e17379fb58579d3b41a007e3783f68cf523a9b1fbbbab2e40c9aa8049f4e2a609c4adc528b904792121e139bbf340980adc38ac32f9538790cad55351b46faae28a0c8d5a995d55e254694102901556c9ce648da6803c6480d5ad0e567ace44d88f5d45cc02e55f81b011d2c2188cd6e89aa6a7cffed6a14a8dbb592e4f386b5a7204d6a0d56cb0844c8df76f7697
random2 = 9d87e8e1b4b74803357d0cc36ca668584c2bba571cb38f66fba3a2d03f897fcf392fd2f88184d330ca03e85b7c4531269849864ee6bc61d81501b7a5796377a73aa6c594b9b37f0e7c130acfac926a329d742ca0bf9edbc39d5bb64dffbbd81467be1fab250eaedbeff79bb2a697584a0b14626833d06f92dbdddd6bc12d6caba725d1eae4a899b4fd75b9bbd9ed05cca9f96c7cc41ae236f3c904f0d9ee4c83111e8dc2bf520a30bcfb2afb8943f61cfd4ef9364dab069276b282f831e1abad31048926ca3cccaa7406998620c941b1ded73d623c0659462e13bf68c0ceafbd5a23afb402b2477297e431182d964cd57cc494b26886af5037964467df5b2f0f981a5aaac9a1718b5b00c174adb0d4de383eb46ffb0ecffc4a432822b9af3d21e7b5b10b7e221978ffb47d058f33456518bc647f78d5e3f5e2850957a12f83b9a0c5a8736b509336eced539c9c0faeb37285f90bb12e3499ef9aaae8c5de67ac2e5b374dffe4e1c5c3f14c4cd8905c7a7db03ac69dbdcf627c2bd4f98dda671dd0da9b70495e476fdba17894665ee2c3f8bb9e5d015b0ee727ddf85115d5833fd13633afd2dd4d5bfa751c9a4b7e937c77939cb72f057714ad94a112d28b578ce265306b415339280748fd20977967f5b04ff92d93e994fe3fd4b887bb8fb7c8cd16af9877532e49df4ee3b7132fed5920d5fc5aeda6784f4371af614d1502f036c1519c53bec2a79303050562b1b929cf705254607486c550fbb40fb7445262929bb4ef8ed658c9e4a825ca0d3e796e5e681a5ce19c2c8880a44c78f205add4c8092a790f7a8d2fbae663d32ac0640317a9735e43369ed680247789330a372910a7ce356f8eafd8776972315bb83794589f22f21770cb1b7c17f25861da3fd142398a85ba357865da0684e916eade4d078deb78ccd5442c27d4e4973eb582d2d55fd5877c64f234a61dcc11a828afa5a52ddea02ad8001c09a60fc7a14bd7118486722db6bf3a54dded4691a7937c3cb546395fc99e49954216f203aaee05985a25ea284022c72dd0a156f030c6a2604b9332ab5324c35f77674836321bcd263f8c2176558d83e23cc00d0fa4d438e9536ab7e574067baeeaeda20b2af5f8ee0821e5ff7949a50c286da61b19d8c32cca7ce1459989731cedd2852a297a6b65637413747b47286a62cf4eabfb5308acbf2d593769f0e8d148945556f1fcfdc7cadd8a6e0c0e174ddb301610d201c7b78b079c2270ac5737fe4771d9bab5257dcf5cca173af2f374c6f68e4658189f1aff3bb81c99e56dc69d14249b07d36c2f73adcb5fa1b71b629a0957b571bf6af13c12983056b6263c9c484b0778c3c1eb1261399644e942eb54479461af9397d7413822a4f267304b1dafc32744ef543a3f3442e05d70c727ac1afb846454aa50d2d979349c508f46981c297672ee78a1
